core/src/mender-utils.c

```diff
@@ -509,3 +509,200 @@ mender_utils_append_list(mender_key_value_list_t **list1, mender_key_value_list_
+    mender_key_value_list_t *item    = *list;
+    while (NULL != item) {
+        if (0 == strcmp(item->key, key)) {
+            to_free = item;
+            if (NULL == prev) {
+                *list = item->next;
+                break;
+            } else {
+                prev->next = item->next;
+            }
+            break;
+        }
+        prev = item;
+        item = item->next;
+    }
+
+    free(to_free->key);
+    free(to_free->value);
+    free(to_free);
+    return MENDER_OK;
+}
+
+mender_err_t
+mender_utils_compare_wildcard(const char *str, const char *wildcard_str, bool *match) {
+
+    assert(NULL != str);
+    assert(NULL != wildcard_str);
+    assert(NULL != match);
+
+    size_t       substrs_pos    = 0;
+    char       **substrs        = NULL;
+    char       **tmp_substrs    = NULL;
+    size_t       escaped_pos    = 0;
+    bool         found_wildcard = false;
+    mender_err_t ret            = MENDER_FAIL;
+
+    *match = true;
+
+    char *escaped = malloc(strlen(wildcard_str) + 1);
+    if (NULL == escaped) {
+        mender_log_error("Unable to allocate memory for escaped str");
+        return MENDER_FAIL;
+    }
+
+    const size_t wildcard_str_len = strlen(wildcard_str);
+
+    /* Split str into substrs on each wildcard */
+    for (size_t i = 0; i < wildcard_str_len; i++) {
+        if ('*' == wildcard_str[i]) {
+            found_wildcard       = true;
+            escaped[escaped_pos] = '\0';
+
+            /* Allocate memory for substrs
+             * temporarily assign it to tmp_substrs in case of NULL*/
+            tmp_substrs = realloc(substrs, (substrs_pos + 1) * sizeof(char *));
+            if (NULL == tmp_substrs) {
+                mender_log_error("Unable to allocate memory");
+                *match = false;
+                goto END;
+            }
+            substrs = tmp_substrs;
+
+            substrs[substrs_pos] = strdup(escaped);
+            if (NULL == substrs[substrs_pos]) {
+                mender_log_error("Unable to allocate memory");
+                *match = false;
+                goto END;
+            }
+
+            substrs_pos++;
+            escaped_pos = 0;
+        } else {
+            escaped[escaped_pos++] = wildcard_str[i];
+        }
+    }
+
+    escaped[escaped_pos] = '\0';
+    substrs              = realloc(substrs, (substrs_pos + 1) * sizeof(char *));
+    if (NULL == substrs) {
+        mender_log_error("Unable to allocate memory");
+        *match = false;
+        goto END;
+    }
+
+    substrs[substrs_pos] = strdup(escaped);
+    if (NULL == substrs[substrs_pos]) {
+        mender_log_error("Unable to allocate memory");
+        *match = false;
+        goto END;
+    }
+    substrs_pos++;
+
+    ret = MENDER_OK;
+
+    /* If no wildcard was found, perform a direct str comparison */
+    if (!found_wildcard) {
+        if (0 != strcmp(str, wildcard_str)) {
+            *match = false;
+            goto END;
+        }
```

Comment:
What if no wildcard was found, but the strings are equal?
```suggestion
        *match = (0 == strcmp(str, wildcard_str));
        goto END;
```

Reply:
Good catch! I forgot to re-base some changes to my unit-test branch so I didn't catch this